ee/agent/reset.go

```diff
@@ -0,0 +1,344 @@
+		if err := k.HostDataStore().Set(hostDataKeySerial, []byte(currentSerial)); err != nil {
+			k.Slogger().Log(ctx, slog.LevelWarn, "could not set serial in host data store", "err", err)
+		}
+		if err := k.HostDataStore().Set(hostDataKeyHardwareUuid, []byte(currentHardwareUUID)); err != nil {
+			k.Slogger().Log(ctx, slog.LevelWarn, "could not set hardware UUID in host data store", "err", err)
+		}
+		if err := k.HostDataStore().Set(hostDataKeyMunemo, []byte(currentTenantMunemo)); err != nil {
+			k.Slogger().Log(ctx, slog.LevelWarn, "could not set munemo in host data store", "err", err)
+		}
+	}
+}
+
+// currentSerialAndHardwareUUID queries osquery for the required information.
+func currentSerialAndHardwareUUID(ctx context.Context, k types.Knapsack) (string, string, error) {
+	osqPath := k.LatestOsquerydPath(ctx)
+	if osqPath == "" {
+		return "", "", errors.New("could not get osqueryd path from knapsack")
+	}
+
+	query := `
+	SELECT
+		system_info.hardware_serial,
+		system_info.uuid AS hardware_uuid
+	FROM
+		system_info;
+`
+
+	var respBytes bytes.Buffer
+	var stderrBytes bytes.Buffer
+
+	osqProc, err := runsimple.NewOsqueryProcess(osqPath, runsimple.WithStdout(&respBytes), runsimple.WithStderr(&stderrBytes))
+	if err != nil {
+		return "", "", fmt.Errorf("could not create osquery process to determine hardware UUID or serial: %w", err)
+	}
+
+	osqCtx, osqCancel := context.WithTimeout(ctx, 10*time.Second)
+	defer osqCancel()
+
+	if sqlErr := osqProc.RunSql(osqCtx, []byte(query)); osqCtx.Err() != nil {
+		return "", "", fmt.Errorf(
+			"querying hardware UUID and serial returned ctx error: `%w` with stdout `%s` and stderr `%s`",
+			osqCtx.Err(), respBytes.String(), stderrBytes.String(),
+		)
+	} else if sqlErr != nil {
+		return "", "", fmt.Errorf(
+			"querying hardware UUID and serial returned error: `%w` with stdout `%s` and stderr `%s`",
+			sqlErr, respBytes.String(), stderrBytes.String(),
+		)
+	}
+
+	var resp []map[string]string
+	if err := json.Unmarshal(respBytes.Bytes(), &resp); err != nil {
+		return "", "", fmt.Errorf("unmarshalling response: %w", err)
+	}
+
+	if len(resp) < 1 {
+		return "", "", errors.New("no rows returned")
+	}
+
+	serial, ok := resp[0]["hardware_serial"]
+	if !ok {
+		return "", "", errors.New("hardware_serial missing from results")
+	}
+
+	hardwareUUID, ok := resp[0]["hardware_uuid"]
+	if !ok {
+		return "", "", errors.New("hardware_uuid missing from results")
+	}
+
+	return serial, hardwareUUID, nil
+}
+
+// valueChanged checks the knapsack for the given data and compares it to
+// currentValue. A value is considered changed only if the stored value was
+// previously set.
+func valueChanged(k types.Knapsack, currentValue string, dataKey []byte) bool {
+	storedValue, err := k.HostDataStore().Get(dataKey)
+	if err != nil {
+		k.Slogger().Log(context.TODO(), slog.LevelError, "could not get stored value", "err", err, "key", string(dataKey))
+		return false // assume no change
+	}
+
+	if storedValue == nil {
```

Comment:
is it possible this will be an empty []byte?
```suggestion
	if storedValue == nil || len(storedValue) == 0 {
```
maybe?

Reply:
Probably don't need the `== nil` then. `len(...) == 0 ` covers it

Reply:
It's nil if the key isn't set -- I think it could be empty if we've stored an empty value previously. I've updated here + the other place where I had a similar check to `len(...) == 0` 👍